File: public/js/core/modules/module-registry.js
// module-registry.js — module registry: registration is open by
// design (no permission, no gatekeeping to publish — anyone can
// register any module at any time), but registration still validates
// two things, both already backed by real code elsewhere in this
// project rather than asserted in prose:
//
//   1. Economic self-declaration: a module that issues accrual must
//      declare whether its reward is time-sensitive (the runtime
//      derives the strong/weak identity scheme from this
//      automatically) and its economic config (α, identity-cost
//      mechanism, scarcity policy). This is enforced here, not just
//      described: α ≤ 1 with no identity-cost mechanism has an
//      unbounded splitting incentive, and is rejected.
//   2. Content integrity (module-hash.js): a registration binds an id
//      to a specific code hash, not a mutable URL.
//
// What this file deliberately does NOT do: decide whether a module's
// CODE is safe, malicious, or good. That's the audit step — per the
// project's own stated direction, an AI-driven audit layered on top of
// this registry later, not a rule enforced at registration time. This
// registry's job is narrower and more mechanical: make sure a judgment,
// whenever and however it's made, has something durable to attach to.

/**
 * @typedef {'strong' | 'weak'} IdentityScheme
 * @typedef {'unaudited' | 'passed' | 'red-listed'} AuditStatus
 *
 * @typedef {{
 *   alpha: number,
 *   identityCostMechanism: string | null,
 *   scarcityPolicy: string,
 * }} EconomicConfig
 *
 * @typedef {{
 *   id: string,
 *   name: string,
 *   icon: string,
 *   category: string,
 *   description: string,
 *   codeHash: string,
 *   codeUrl: string,
 *   author: string,
 *   isIssuing: boolean,
 *   timeSensitive: boolean | null,
 *   economicConfig: EconomicConfig | null,
 *   identityScheme: IdentityScheme | null,
 *   auditStatus: AuditStatus,
 *   registeredAt: number,
 * }} ModuleEntry
 *
 * @typedef {{ modules: Record<string, ModuleEntry> }} ModuleRegistryState
 */

export function initialModuleRegistryState() {
  return { modules: {} };
}

/**
 * The runtime does not ask a module author to pick an identity scheme
 * — it derives the minimum sufficient one from the one fact that
 * actually determines it: whether the module's reward function is
 * sensitive to elapsed cadence time. A module that declares itself
 * time-insensitive gets the cheaper weak scheme automatically; nothing
 * stronger is forced on it, and nothing weaker is permitted if it
 * declared time-sensitivity.
 *
 * @param {boolean} timeSensitive
 * @returns {IdentityScheme}
 */
export function selectIdentityScheme(timeSensitive) {
  return timeSensitive ? 'strong' : 'weak';
}

/**
 * Enforced rather than only described: with reward concave or
 * linear in committed capital (α ≤ 1) and no identity-cost mechanism,
 * splitting into many identities is non-decreasing in profit and
 * reward concavity alone gives no bound — a module configured this way
 * has, by the paper's own algebra, an unbounded Sybil-splitting
 * incentive. This is the one case registration actually refuses, not
 * because the code is judged unsafe, but because the declared economics
 * are internally inconsistent with what this project's own analysis
 * requires.
 *
 * @param {EconomicConfig} config
 * @returns {{ valid: boolean, reason?: string }}
 */
export function validateEconomicConfig(config) {
  if (!Number.isFinite(config.alpha)) {
    return { valid: false, reason: 'alpha must be a finite number' };
  }
  if (config.alpha <= 1 && !config.identityCostMechanism) {
    return {
      valid: false,
      reason: 'alpha <= 1 with no identity-cost mechanism has an unbounded splitting incentive — declare an identityCostMechanism or raise alpha above 1',
    };
  }
  if (!config.scarcityPolicy) {
    return { valid: false, reason: 'scarcityPolicy must be declared' };
  }
  return { valid: true };
}

/**
 * Registers a module. Always open — no author allow-list, no approval
 * step, no minimum reputation to publish. The only rejections are
 * mechanical: a duplicate id, or (for an issuing module only) an
 * internally-inconsistent economic declaration per validateEconomicConfig.
 * A non-issuing module (isIssuing: false) skips economic validation
 * entirely and registers in read-only mode.
 *
 * @param {ModuleRegistryState} state
 * @param {Omit<ModuleEntry, 'identityScheme' | 'auditStatus' | 'registeredAt'>} entry
 * @param {{ now?: number }} [opts]
 * @returns {{ state: ModuleRegistryState, accepted: boolean, reason?: string }}
 */
export function registerModule(state, entry, { now = Date.now() } = {}) {
  if (state.modules[entry.id]) {
    return { state, accepted: false, reason: `module id '${entry.id}' is already registered` };
  }

  let identityScheme = null;
  if (entry.isIssuing) {
    if (typeof entry.timeSensitive !== 'boolean') {
      return { state, accepted: false, reason: 'an issuing module must declare timeSensitive (true or false)' };
    }
    if (!entry.economicConfig) {
      return { state, accepted: false, reason: 'an issuing module must declare economicConfig' };
    }
    const check = validateEconomicConfig(entry.economicConfig);
    if (!check.valid) {
      return { state, accepted: false, reason: check.reason };
    }
    identityScheme = selectIdentityScheme(entry.timeSensitive);
  }

  const record = { ...entry, identityScheme, auditStatus: 'unaudited', registeredAt: now };
  return { state: { modules: { ...state.modules, [entry.id]: record } }, accepted: true };
}

/**
 * Publishes a new code version for an already-registered module —
 * still open, no permission required, matching the project's stated
 * "plug and submit, as frictionless as possible" direction. What
 * changes going forward (module-hash.js) is that this is a distinct,
 * recorded event with its own hash: any judgment already made about
 * the previous codeHash does not silently carry over to this one.
 *
 * @param {ModuleRegistryState} state
 * @param {{ id: string, codeHash: string, codeUrl: string }} params
 */
export function updateModuleCode(state, { id, codeHash, codeUrl }) {
  const existing = state.modules[id];
  if (!existing) {
    return { state, accepted: false, reason: `module id '${id}' is not registered` };
  }
  const updated = { ...existing, codeHash, codeUrl, auditStatus: 'unaudited' };
  return { state: { modules: { ...state.modules, [id]: updated } }, accepted: true };
}

/**
 * Records an audit verdict against a module's CURRENT codeHash. Not an
 * audit implementation — a hook for one (per the project's stated
 * direction, an AI-driven audit, built later) to attach a durable
 * verdict to. Because updateModuleCode() resets auditStatus to
 * 'unaudited' on every code change, a verdict can never silently apply
 * to a version of the code it was never actually about.
 *
 * @param {ModuleRegistryState} state
 * @param {string} id
 * @param {import('./module-registry.js').AuditStatus} status
 */
const VALID_AUDIT_STATUSES = ['unaudited', 'passed', 'red-listed'];

/**
 * SECURITY/PARITY: previously accepted any string as `status` with no
 * validation at all (enum membership was only enforced in a JSDoc
 * comment) — Rust's own set_audit_status() already required a real
 * AuditStatus enum member. Given an identical 'module-audit' event,
 * the two languages could materialize different auditStatus values,
 * a real cross-language divergence found by direct inspection, not
 * hypothetical. Fixed to validate the same closed set Rust's own enum
 * defines.
 */
export function setAuditStatus(state, id, status) {
  const existing = state.modules[id];
  if (!existing) {
    return { state, accepted: false, reason: `module id '${id}' is not registered` };
  }
  if (!VALID_AUDIT_STATUSES.includes(status)) {
    return { state, accepted: false, reason: `'${status}' is not a valid audit status — must be one of ${VALID_AUDIT_STATUSES.join(', ')}` };
  }
  return { state: { modules: { ...state.modules, [id]: { ...existing, auditStatus: status } } }, accepted: true };
}

File: rust-core/src/modules/module_registry.rs
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum IdentityScheme {
    Strong,
    Weak,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum AuditStatus {
    Unaudited,
    Passed,
    RedListed,
}

#[derive(Debug, Clone)]
pub struct EconomicConfig {
    pub alpha: f64,
    pub identity_cost_mechanism: Option<String>,
    pub scarcity_policy: String,
}

#[derive(Debug, Clone)]
pub struct ModuleEntry {
    pub id: String,
    pub name: String,
    pub icon: String,
    pub category: String,
    pub description: String,
    pub code_hash: String,
    pub code_url: String,
    pub author: String,
    pub is_issuing: bool,
    pub time_sensitive: Option<bool>,
    pub economic_config: Option<EconomicConfig>,
    pub identity_scheme: Option<IdentityScheme>,
    pub audit_status: AuditStatus,
    pub registered_at: i64,
}

#[derive(Debug, Clone, Default)]
pub struct ModuleRegistryState {
    pub modules: HashMap<String, ModuleEntry>,
}

/// §11/§27.2: derives the minimum sufficient identity scheme from
/// whether the module's reward function is time-sensitive — mirror of
/// selectIdentityScheme() in module-registry.js.
pub fn select_identity_scheme(time_sensitive: bool) -> IdentityScheme {
    if time_sensitive {
        IdentityScheme::Strong
    } else {
        IdentityScheme::Weak
    }
}

#[derive(Debug, PartialEq)]
pub struct ValidationResult {
    pub valid: bool,
    pub reason: Option<String>,
}

/// §24.1, enforced: α ≤ 1 with no identity-cost mechanism has an
/// unbounded splitting incentive — mirror of validateEconomicConfig()
/// in module-registry.js.
pub fn validate_economic_config(config: &EconomicConfig) -> ValidationResult {
    if !config.alpha.is_finite() {
        return ValidationResult { valid: false, reason: Some("alpha must be a finite number".to_string()) };
    }
    if config.alpha <= 1.0 && config.identity_cost_mechanism.is_none() {
        return ValidationResult {
            valid: false,
            reason: Some(
                "alpha <= 1 with no identity-cost mechanism has an unbounded splitting incentive (§24.1) — declare an identity_cost_mechanism or raise alpha above 1"
                    .to_string(),
            ),
        };
    }
    if config.scarcity_policy.is_empty() {
        return ValidationResult { valid: false, reason: Some("scarcity_policy must be declared".to_string()) };
    }
    ValidationResult { valid: true, reason: None }
}

pub struct NewModule {
    pub id: String,
    pub name: String,
    pub icon: String,
    pub category: String,
    pub description: String,
    pub code_hash: String,
    pub code_url: String,
    pub author: String,
    pub is_issuing: bool,
    pub time_sensitive: Option<bool>,
    pub economic_config: Option<EconomicConfig>,
}

#[derive(Debug, PartialEq)]
pub struct RegisterOutcome {
    pub accepted: bool,
    pub reason: Option<String>,
}

impl ModuleRegistryState {
    pub fn new() -> Self {
        Self::default()
    }

    /// Registration is open — no author allow-list, no approval step.
    /// The only rejections are mechanical: a duplicate id, or (for an
    /// issuing module only) an internally-inconsistent economic
    /// declaration. Mirror of registerModule() in module-registry.js.
    pub fn register_module(&mut self, entry: NewModule, now: i64) -> RegisterOutcome {
        if self.modules.contains_key(&entry.id) {
            return RegisterOutcome {
                accepted: false,
                reason: Some(format!("module id '{}' is already registered", entry.id)),
            };
        }

        let mut identity_scheme = None;
        if entry.is_issuing {
            let time_sensitive = match entry.time_sensitive {
                Some(v) => v,
                None => {
                    return RegisterOutcome {
                        accepted: false,
                        reason: Some("an issuing module must declare time_sensitive".to_string()),
                    }
                }
            };
            let config = match &entry.economic_config {
                Some(c) => c,
                None => {
                    return RegisterOutcome {
                        accepted: false,
                        reason: Some("an issuing module must declare economic_config".to_string()),
                    }
                }
            };
            let check = validate_economic_config(config);
            if !check.valid {
                return RegisterOutcome { accepted: false, reason: check.reason };
            }
            identity_scheme = Some(select_identity_scheme(time_sensitive));
        }

        self.modules.insert(
            entry.id.clone(),
            ModuleEntry {
                id: entry.id,
                name: entry.name,
                icon: entry.icon,
                category: entry.category,
                description: entry.description,
                code_hash: entry.code_hash,
                code_url: entry.code_url,
                author: entry.author,
                is_issuing: entry.is_issuing,
                time_sensitive: entry.time_sensitive,
                economic_config: entry.economic_config,
                identity_scheme,
                audit_status: AuditStatus::Unaudited,
                registered_at: now,
            },
        );
        RegisterOutcome { accepted: true, reason: None }
    }

    /// Still open, no permission required — resets audit_status to
    /// Unaudited so a stale verdict never silently carries over to new
    /// code. Mirror of updateModuleCode().
    pub fn update_module_code(&mut self, id: &str, code_hash: String, code_url: String) -> RegisterOutcome {
        match self.modules.get_mut(id) {
            Some(m) => {
                m.code_hash = code_hash;
                m.code_url = code_url;
                m.audit_status = AuditStatus::Unaudited;
                RegisterOutcome { accepted: true, reason: None }
            }
            None => RegisterOutcome { accepted: false, reason: Some(format!("module id '{id}' is not registered")) },
        }
    }

    /// A hook for a future (AI-driven, per the project's stated
    /// direction) audit to attach a durable verdict to — not an audit
    /// implementation itself. Mirror of setAuditStatus().
    pub fn set_audit_status(&mut self, id: &str, status: AuditStatus) -> RegisterOutcome {
        match self.modules.get_mut(id) {
            Some(m) => {
                m.audit_status = status;
                RegisterOutcome { accepted: true, reason: None }
            }
            None => RegisterOutcome { accepted: false, reason: Some(format!("module id '{id}' is not registered")) },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base_entry(id: &str) -> NewModule {
        NewModule {
            id: id.to_string(),
            name: "My Module".to_string(),
            icon: "🔮".to_string(),
            category: "Tools".to_string(),
            description: "Does a thing.".to_string(),
            code_hash: "hash1".to_string(),
            code_url: "https://example.com/mymodule.js".to_string(),
            author: "alice".to_string(),
            is_issuing: false,
            time_sensitive: None,
            economic_config: None,
        }
    }

    #[test]
    fn select_identity_scheme_derives_from_time_sensitivity() {
        assert_eq!(select_identity_scheme(true), IdentityScheme::Strong);
        assert_eq!(select_identity_scheme(false), IdentityScheme::Weak);
    }

    #[test]
    fn validate_rejects_alpha_leq_1_with_no_identity_cost() {
        let cfg = EconomicConfig { alpha: 1.0, identity_cost_mechanism: None, scarcity_policy: "preallocated".to_string() };
        assert!(!validate_economic_config(&cfg).valid);
    }

    #[test]
    fn validate_accepts_alpha_leq_1_with_identity_cost() {
        let cfg = EconomicConfig {
            alpha: 0.5,
            identity_cost_mechanism: Some("sol-burn".to_string()),
            scarcity_policy: "preallocated".to_string(),
        };
        assert!(validate_economic_config(&cfg).valid);
    }

    #[test]
    fn validate_accepts_alpha_gt_1_with_no_identity_cost() {
        let cfg = EconomicConfig { alpha: 1.5, identity_cost_mechanism: None, scarcity_policy: "preallocated".to_string() };
        assert!(validate_economic_config(&cfg).valid);
    }

    #[test]
    fn non_issuing_module_registers_without_economic_declaration() {
        let mut state = ModuleRegistryState::new();
        let outcome = state.register_module(base_entry("mymodule.js"), 1000);
        assert!(outcome.accepted);
        assert_eq!(state.modules["mymodule.js"].identity_scheme, None);
        assert_eq!(state.modules["mymodule.js"].audit_status, AuditStatus::Unaudited);
    }

    #[test]
    fn issuing_module_without_economic_config_is_rejected() {
        let mut state = ModuleRegistryState::new();
        let mut entry = base_entry("mymodule.js");
        entry.is_issuing = true;
        entry.time_sensitive = Some(true);
        assert!(!state.register_module(entry, 1000).accepted);
    }

    #[test]
    fn valid_issuing_module_gets_identity_scheme_from_lemma_1() {
        let mut state = ModuleRegistryState::new();
        let mut entry = base_entry("mymodule.js");
        entry.is_issuing = true;
        entry.time_sensitive = Some(true);
        entry.economic_config = Some(EconomicConfig { alpha: 2.0, identity_cost_mechanism: None, scarcity_policy: "preallocated".to_string() });
        let outcome = state.register_module(entry, 1000);
        assert!(outcome.accepted);
        assert_eq!(state.modules["mymodule.js"].identity_scheme, Some(IdentityScheme::Strong));
    }

    #[test]
    fn registration_is_open_no_allowlist() {
        let mut state = ModuleRegistryState::new();
        let mut e1 = base_entry("a.js");
        e1.author = "alice".to_string();
        assert!(state.register_module(e1, 1000).accepted);
        let mut e2 = base_entry("b.js");
        e2.author = "someone-nobody-approved".to_string();
        assert!(state.register_module(e2, 1001).accepted);
    }

    #[test]
    fn duplicate_id_rejected() {
        let mut state = ModuleRegistryState::new();
        state.register_module(base_entry("a.js"), 1000);
        let outcome = state.register_module(base_entry("a.js"), 1001);
        assert!(!outcome.accepted);
    }

    #[test]
    fn update_module_code_resets_audit_status() {
        let mut state = ModuleRegistryState::new();
        state.register_module(base_entry("mymodule.js"), 1000);
        state.set_audit_status("mymodule.js", AuditStatus::Passed);
        assert_eq!(state.modules["mymodule.js"].audit_status, AuditStatus::Passed);

        state.update_module_code("mymodule.js", "hash2".to_string(), "https://example.com/m.js".to_string());
        assert_eq!(state.modules["mymodule.js"].audit_status, AuditStatus::Unaudited);
        assert_eq!(state.modules["mymodule.js"].code_hash, "hash2");
    }

    #[test]
    fn set_audit_status_supports_red_listing() {
        let mut state = ModuleRegistryState::new();
        state.register_module(base_entry("mymodule.js"), 1000);
        state.set_audit_status("mymodule.js", AuditStatus::RedListed);
        assert_eq!(state.modules["mymodule.js"].audit_status, AuditStatus::RedListed);
    }
}
